compiler/damlc/tests/daml-test-files/UnusedMatchTests.daml

```diff
@@ -0,0 +1,36 @@
+-- Copyright (c) 2020, Digital Asset (Switzerland) GmbH and/or its
+-- affiliates. All rights reserved.
+
+-- @WARN range=12:2-12:3; Defined but not used
+
+{-# OPTIONS_GHC -Wunused-matches #-}
+module UnusedMatchTests where
+
+-- It should be OK to enable -Wunused-matches and not get warnings
+-- from template desugared code.
+
+f x = 12 -- Defined but not used 'x'; prove unsed match detection.
+
+template T
+  with
+    p : Party
+    q : Party
+  where
+    -- Include interaction with local binds.
+    let revokeRetVal = ()
+        (sig, obs) = (p, q)
```

Comment:
I might be missing something here but I thought your implementation will not try to silence the unused binds here (which I think is totally reasonable, I’ve never seen anyone write this in DAML) so how does CI pass?

Reply:
Those are regular let bindings. They end up in the where clause of an implicit function e.g.
```haskell
  signatory this =
    ...
    where (a, b) = (0, 1)
```
For some reason, GHC doesn't care about them (!) as evidenced by this program:
```haskell
{-# OPTIONS_GHC -Wunused-matches #-}

f x = 23
  where (a, b) = (0, 1)

main = undefined
```
/Users/shaynefletcher/Foo.hs:3:3: warning: [-Wunused-matches]
    Defined but not used: ‘x’
  |
3 | f x = 23
  |   ^
```

Reply:
That's not new by the way - we didn't write code that processes let bindings, only arguments like `this`, `self` and `arg` are getting the "dummy binding" treatment.

Reply:
Oh I see, you didn’t enable `-Wunused-local-binds`. I would suggest to just go for `-Wall` and selectively disable stuff.

Reply:
> That's not new by the way - we didn't write (any new) code that processes let bindings, only arguments like `this`, `self` and `arg` are getting the "dummy binding" treatment.

Oh I thought let bindings worked if they are variables or as-pattern? If not, we should make them work.

Reply:
We pass all arguments to `signatory` and friends to the `dummyBinds` function so all as patterns and var patterns are given that treatment. Bindings in a template `let` block are not "function parameters" and are not given any new treatments. Are you suggesting we should be doing more?

Reply:
> Oh I see, you didn’t enable `-Wunused-local-binds`

Right, only `unused-matches`.

> I would suggest to just go for `-Wall` and selectively disable stuff so we capture all the unused stuff GHC has to offer.

I don't disagree but you're not arguing to extend the scope of this PR are you?

Reply:
> I don't disagree but you're not arguing to extend the scope of this PR are you?

I’m not suggesting to change what we do but I am suggesting to enable more warnings so we have an overview of which ones are solved now and which ones are still pending.

Reply:
> > Oh I see, you didn’t enable `-Wunused-local-binds`
> 
> Right, only `unused-matches`.
> 
Indeed with `unused-local-binds` we get the expected warnings on `a, b`.

Reply:
> > I don't disagree but you're not arguing to extend the scope of this PR are you?
> 
> I’m not suggesting to change what we do but I am suggesting to enable more warnings so we have an overview of which ones are solved now and which ones are still pending.

Sounds good. Yep more to do.